Validate config keys

orkestrov opened the pull request:
Related issue: https://github.com/github/orchestrator/issues/82

### Description

Simple config validation. Parsing JSON twice: firstly into a hash map for validation, secondly into a Config struct. Currently only parameter names are checked.

orkestrov committed:
simple config keys validation

read() function dies on unknown config keys.
Keys prefixed with underscore are ignored.

orkestrov committed:
refactoring

Commit:
Merge branch 'master' into validate_config_keys

sjmudd committed:
Merge branch 'master' into validate_config_keys

Commit:
Merge branch 'master' into validate_config_keys

shlomi-noach commented:
Are you able to add unit tests for:
- confirming valid configuration is not failing the code
- confirming `_xyz` configuration is silently ignored
- confirming invalid configuration is failing the code?

The `log.Fatalf()` makes it hard to test. I suggest that on error `validate()` and `read()` should just return the error, and that the handling code should do the `Fatal` part.

sjmudd committed:
Merge branch 'master' into validate_config_keys

sjmudd commented:
I also think that given I have seen invalid configurations in the past that by default invalid settings should not break things as this may cause issues for people who are not aware of the sudden changes.
* log all the invalid settings in a single line rather than try to be fatal on the first one you find. It's probably better to log all issues rather than the first one that is found.
* Maybe add a setting `ConfigValidation = [ WARN | FATAL ]`
* Check what happens if the configuration is reread e.g. due to receiving SIGHUP as we probably do not want to kill an orchestrator process that's already running (?)
* Maybe even add a orchestrator -c check_config to validate the configuration without actually running. Might be useful?

So let's look to see if we can get these changes in. The validation is useful, let's not make that cause extra trouble for people but allow us to catch things if we want to.

Commit:
Merge branch 'master' into validate_config_keys

orkestrov committed:
couple of config tests and not die on SIGHUP

Read() considers config parsing error as fatal,
Reload() only writes an error to the error log.

orkestrov commented:
I pushed a couple of config tests and also following @sjmudd's advice made sure that orchestrator does not die on SIGHUP when config is erroneous.

sjmudd committed:
Merge branch 'master' into validate_config_keys

sjmudd committed:
Merge branch 'master' into validate_config_keys

Commit:
Merge branch 'master' into validate_config_keys

Commit:
Merge branch 'master' into validate_config_keys

shlomi-noach commented:
@orkestrov haha this is so meta!

So I've just added TravisCI which will run public CI for PRs. As you can see, your PR fails right now.

But why is it failing? It's because the tests use a variable called `AutoRecoveryGloballyEnabled` which is not in the Config.

So your PR just tested itself. Please remove the variable from `tests/integration/orchestrator.conf.json` and let's see if your PR builds.

orkestrov commented:
@shlomi-noach I used to test with `go test ./go/...` which works. I probably missed the moment we started running integration tests. Is there an intro/requirements for this? When I run `./tests/integration/test.sh` I get this:

`ERROR 2002 (HY000): Can't connect to local MySQL server through socket '/var/lib/mysql/mysql.sock' (2)
`

shlomi-noach commented:
@orkestrov like I said, "So I've just added TravisCI which will run public CI for PRs." 

We were having private-only CI and as of now we have public CI.

> Is there an intro/requirements for this?

CI must pass in order to merge a PR.

did you do anything wrong?

No, the CI tests actually _validate_ your PR. And you have the honor of fixing the config file.

orkestrov committed:
rm deprecated config option from tests

orkestrov commented:
@shlomi-noach hmm I don't mind fixing the config as you suggest. What I'm missing is: how can I test it or get Travis logs? I just removed the `AutoRecoveryGloballyEnabled` as you suggested but it does not seem to have fixed the problem.

shlomi-noach commented:
> What I'm missing is: how can I test it or get Travis logs?

If you like to test this locally, I'll walk you through this. However you don't need to, aince TravisCI will do it for you.
Are you not able to see a "Details" link to the right of the failing test? And does it not show you the build/test log?

shlomi-noach commented:
<img width="847" alt="screen shot 2017-04-03 at 14 56 26" src="https://cloud.githubusercontent.com/assets/2607934/24608121/c3431b28-187d-11e7-9215-b80df384bf5e.png">

shlomi-noach commented:
@orkestrov regardless, to be able to run the tests locally, you need a mysql server running on your box, with no credentials.

orkestrov commented:
@shlomi-noach I don't have the "Details" link, but I just figured out that I can click the red cross next to the build to reach the logs. Got it now, thanks for the prompt replies.

shlomi-noach commented:
@orkestrov interesting. For me the red cross is not clickable and I get a "details" link.

orkestrov committed:
rm deprecated config option from tests

shlomi-noach commented:
@orkestrov `OAuthClientID` should be `OAuthClientId`

orkestrov commented:
@shlomi-noach probably we get different results due to some GH frontend experiment:
![image](https://cloud.githubusercontent.com/assets/20597053/24608336/71d996c4-1876-11e7-8936-dce77abf4897.png)

orkestrov committed:
rm deprecated config option from tests

shlomi-noach commented:
@orkestrov `OAuthClientID` should be `OAuthClientId` -- it's just capitalization

orkestrov committed:
add OAuthClientId back fixing the field name

shlomi-noach commented:
woohoo!

orkestrov commented:
> @orkestrov OAuthClientID should be OAuthClientId -- it's just capitalisation

Thanks for spotting, added it back now.

orkestrov commented:
Cool! it worked!

Nice to have the CI btw. This branch has been very meta indeed :-)

On Mon, Apr 3, 2017 at 2:13 PM, Shlomi Noach <notifications@github.com>
wrote:

> woohoo!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/github/orchestrator/pull/95#issuecomment-291125316>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ATpJPeuZSaiF1N-0cFJJNfI7-6YqlHOtks5rsOJggaJpZM4MW1qG>
> .
>

Commit:
Merge branch 'master' into validate_config_keys
